### PR TITLE
AirfRANS: Re-stratified sampling for Reynolds number OOD robustness

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Re-stratified sampling (--re-stratified-sampling) ensures balanced Reynolds number (Re) representation across training batches. AirfRANS spans a wide Re range and random mini-batch sampling can under-represent low/high-Re regimes, biasing the model toward the mean-Re distribution. By ensuring each batch covers the Re spectrum uniformly, we expect improved OOD robustness — particularly on the scarce/extreme-Re validation cases that drive surface_mse.

This flag comes from the noam branch stack, but has never been tested in isolation on AF. Multiple in-flight PRs (#3184 wolfwood, #3187 stark, #3191 alphonse, #3177 nezuko) all test aspects of the noam stack bundled together. This PR isolates re-stratified sampling alone to determine its individual contribution.

## Instructions

Run two trials sequentially. Start with a 3-epoch smoke test on Trial 1 to verify training is stable, then report val_primary/surface_mse and val_primary/vol_mse.

**Trial 1 — Re-stratified on AF base champion (no vol-weight):**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --optimizer adamw \
  --lr 6e-4 \
  --cosine-t-max 50 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 \
  --model-hidden-dim 256 \
  --model-heads 8 \
  --epochs 999 \
  --ema-decay 0.999 \
  --re-stratified-sampling \
  --wandb_group af-re-stratified
```

**Trial 2 — Re-stratified on AF vol-weight=10x champion:**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --optimizer adamw \
  --lr 6e-4 \
  --cosine-t-max 50 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 \
  --model-hidden-dim 256 \
  --model-heads 8 \
  --epochs 999 \
  --ema-decay 0.999 \
  --re-stratified-sampling \
  --vol-loss-weight 10.0 \
  --wandb_group af-re-stratified
```

Note: --model-heads 8 (vs baseline --model-heads 4). If --model-heads 8 is not supported or causes issues, revert to --model-heads 4 to match the exact baseline config.

**Smoke test:** Run Trial 1 for 3 epochs first. Confirm training is stable (no NaN loss) then proceed to full runs.

## Baseline

- **val_primary/surface_mse:** 0.000296 (epoch 704) — beat this
- **val_primary/vol_mse:** 0.002039 (epoch 743) — beat this
- **Best PR:** #3135 (nezuko — EMA=0.999 + vol-weight=10x, 2L/256d/4H, AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, Fourier)
- **W&B run:** sh2zyfwr (nezuko/af-ema-vol-weight-10x)
- **Reproduce baseline:** cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0

## Expected outcome

Re-stratified sampling may improve OOD generalization across Re regimes. Possible outcomes:
- Improves surface_mse (primary win): Re-balanced batches reduce overfitting to mean-Re cases
- Improves vol_mse (secondary win): Reynolds-aware sampling may regularize volume predictions too
- No effect or regression: If the model already generalizes well across Re, the rebalancing may hurt by under-sampling the dominant regime

Report best val_primary/surface_mse and val_primary/vol_mse for each trial, and the epoch each best was achieved.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT piccolo:

## Results

### Trial 1 — Re-stratified sampling, no vol-weight (8H)

| Metric | Baseline | Trial 1 | Delta |
|--------|----------|---------|-------|
| val_primary/surface_mse | 0.000296 (ep704) | **0.000582** (ep613) | +96.6% ❌ |
| val_primary/vol_mse | 0.002039 (ep743) | 0.003570 (ep613) | +75.1% ❌ |
| Best vol_mse | — | 0.003425 (ep589) | — |

- **Test metrics (best checkpoint, ep613):** test surface_mse=0.001323, test vol_mse=0.004196
- **Total epochs:** 643 (360-min timeout)
- **W&B run:** f2g4mhvk (`piccolo/af-re-strat-trial1-no-vol`)
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 8 --epochs 999 --ema-decay 0.999 --re-stratified-sampling --agent piccolo --wandb-name "piccolo/af-re-strat-trial1-no-vol" --wandb-group af-re-stratified
```

### Trial 2 — Re-stratified sampling + vol-weight=10x (8H)

| Metric | Baseline | Trial 2 | Delta |
|--------|----------|---------|-------|
| val_primary/surface_mse | 0.000296 (ep704) | **0.000608** (ep452) | +105.4% ❌ |
| val_primary/vol_mse | 0.002039 (ep743) | 0.003831 (ep452) | +87.9% ❌ |
| Best vol_mse | — | 0.003737 (ep442) | — |

- **Test metrics (best checkpoint, ep452):** test surface_mse=0.001089, test vol_mse=0.005215
- **Total epochs:** 653 (360-min timeout), **NaN divergence at epoch 484**
- **W&B run:** nh3i11z4 (`piccolo/af-re-strat-trial2-vol10x`)
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 8 --epochs 999 --ema-decay 0.999 --re-stratified-sampling --volume-loss-weight 10.0 --agent piccolo --wandb-name "piccolo/af-re-strat-trial2-vol10x" --wandb-group af-re-stratified
```

## What happened

**Re-stratified sampling did not improve either metric.** Both trials are substantially worse than the baseline across surface_mse (~2x) and vol_mse (~1.8x).

Key observations:
1. **Trial 1** (no vol-weight) trained stably through 643 epochs but never approached the baseline 0.000296. Best at ep613, still ~2x worse.
2. **Trial 2** (vol-weight=10x) diverged to NaN at epoch 484 — the re-stratified sampling likely introduced gradient instability that the vol-weight amplified. EMA preserved the best checkpoint (ep452) but the model never recovered.
3. **Confound: 8 heads vs 4 heads.** Both trials used `--model-heads 8` per PR instructions (baseline uses 4H). 256d/8H = 32 per head vs 256d/4H = 64 per head — the smaller head dimension may have reduced the model's capacity to capture fine-grained pressure fields, compounding any effect from re-stratified sampling.
4. **Hypothesis likely falsified.** The AirfRANS model already generalizes well across Re regimes with random sampling. Re-balancing toward low-frequency Re bins likely under-samples the dominant regime where most signal lives, degrading overall accuracy.

## Suggested follow-ups

- **Isolate the 8H effect:** Run the baseline config with `--model-heads 8` (no re-stratified) to determine how much of the regression is from head count vs sampling strategy.
- **Softer re-stratification:** Instead of inverse-frequency weighting (hard re-balancing), try a sqrt-inverse-frequency or temperature-scaled version that partially corrects the Re distribution without fully flattening it.
- **Re-stratified with 4H:** If 8H is confirmed harmful, re-test re-stratified sampling with the baseline 4H config to isolate the sampling effect.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3195 because ADVISOR: Closing — re-stratified sampling surface=0.000582 (~2x worse), vol=0.003425 (+68%). Reynolds number stratification harms convergence. Trial 2 diverged entirely. Piccolo reassigned..
